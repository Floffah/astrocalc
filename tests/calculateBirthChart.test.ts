import { describe, expect, test } from "bun:test";

import type { CalculateBirthChartResponse } from "@/defs/responses.ts";
import app from "@/index.ts";

describe("Valid", () => {
    test("Call calculateBirthChart with valid parameters", async () => {
        const params = new URLSearchParams();

        params.set("year", "2000");
        params.set("month", "1");
        params.set("day", "1");
        params.set("hour", "5");
        params.set("minute", "30");

        params.set("latitude", "50.123456");
        params.set("longitude", "3.123456");

        const res = await app.request("/birth-chart?" + params.toString());
        const body = (await res.json()) as {
            data: CalculateBirthChartResponse;
        };

        expect(body.data).toBeObject();
        expect(body.data.signs).toMatchObject({
            ascendant: {
                value: "Sagittarius",
                cuspWarning: null,
            },
            moon: {
                value: "Scorpio",
                cuspWarning: null,
                isVoidOfCourse: false,
                phase: "Waxing Crescent",
            },
            sun: {
                value: "Capricorn",
                cuspWarning: null,
            },
        });
        expect(
            body.data.angles.map(
                ({ degree: _, longitude: _1, ...angle }) => angle,
            ),
        ).toMatchSnapshot();

        expect(
            body.data.houses.map(
                ({
                    cusp: { degree: _, longitude: _1, ...cusp },
                    ...house
                }) => ({
                    ...house,
                    cusp,
                }),
            ),
        ).toMatchSnapshot();

        expect(
            body.data.planets.map(
                ({ degree: _, latitude: _1, longitude: _2, ...planet }) =>
                    planet,
            ),
        ).toMatchSnapshot();
    });
});

describe("Invalid", () => {
    test("Call calculateBirthChart with missing parameters", async () => {
        const params = new URLSearchParams();

        params.set("year", "2000");
        params.set("month", "1");
        params.set("day", "1");
        params.set("hour", "5");
        params.set("minute", "30");

        const res = await app.request("/birth-chart?" + params.toString());

        expect(await res.json()).toMatchSnapshot();
    });

    test("Call calculateBirthChart with invalid parameters", async () => {
        const params = new URLSearchParams();

        params.set("year", "2200");
        params.set("month", "18");
        params.set("day", "-5");
        params.set("hour", "24");
        params.set("minute", "30");

        params.set("latitude", "50.123456");
        params.set("longitude", "3.123456");

        const res = await app.request("/birth-chart?" + params.toString());

        expect(await res.json()).toMatchSnapshot();
    });
});
