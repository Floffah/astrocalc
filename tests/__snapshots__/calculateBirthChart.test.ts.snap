// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Valid Call calculateBirthChart with valid parameters 1`] = `
[
  {
    "houseNumber": 1,
    "id": 100,
    "isRetrograde": false,
    "name": "Ascendant",
    "zodiac": {
      "id": 8,
      "lord": {
        "id": 5,
        "name": "Jupiter",
      },
      "name": "Sagittarius",
    },
  },
  {
    "houseNumber": 10,
    "id": 109,
    "isRetrograde": false,
    "name": "MidHeaven",
    "zodiac": {
      "id": 6,
      "lord": {
        "id": 2,
        "name": "Venus",
      },
      "name": "Libra",
    },
  },
  {
    "houseNumber": 7,
    "id": 108,
    "isRetrograde": false,
    "name": "Descendant",
    "zodiac": {
      "id": 2,
      "lord": {
        "id": 3,
        "name": "Mercury",
      },
      "name": "Gemini",
    },
  },
  {
    "houseNumber": 4,
    "id": 107,
    "isRetrograde": false,
    "name": "Nadir",
    "zodiac": {
      "id": 0,
      "lord": {
        "id": 4,
        "name": "Mars",
      },
      "name": "Aries",
    },
  },
]
`;

exports[`Valid Call calculateBirthChart with valid parameters 2`] = `
[
  {
    "cusp": {
      "zodiac": {
        "id": 8,
        "lord": {
          "id": 5,
          "name": "Jupiter",
        },
        "name": "Sagittarius",
      },
    },
    "id": 0,
    "number": 1,
  },
  {
    "cusp": {
      "zodiac": {
        "id": 9,
        "lord": {
          "id": 6,
          "name": "Saturn",
        },
        "name": "Capricorn",
      },
    },
    "id": 1,
    "number": 2,
  },
  {
    "cusp": {
      "zodiac": {
        "id": 10,
        "lord": {
          "id": 6,
          "name": "Saturn",
        },
        "name": "Aquarius",
      },
    },
    "id": 2,
    "number": 3,
  },
  {
    "cusp": {
      "zodiac": {
        "id": 0,
        "lord": {
          "id": 4,
          "name": "Mars",
        },
        "name": "Aries",
      },
    },
    "id": 3,
    "number": 4,
  },
  {
    "cusp": {
      "zodiac": {
        "id": 1,
        "lord": {
          "id": 2,
          "name": "Venus",
        },
        "name": "Taurus",
      },
    },
    "id": 4,
    "number": 5,
  },
  {
    "cusp": {
      "zodiac": {
        "id": 1,
        "lord": {
          "id": 2,
          "name": "Venus",
        },
        "name": "Taurus",
      },
    },
    "id": 5,
    "number": 6,
  },
  {
    "cusp": {
      "zodiac": {
        "id": 2,
        "lord": {
          "id": 3,
          "name": "Mercury",
        },
        "name": "Gemini",
      },
    },
    "id": 6,
    "number": 7,
  },
  {
    "cusp": {
      "zodiac": {
        "id": 3,
        "lord": {
          "id": 1,
          "name": "Moon",
        },
        "name": "Cancer",
      },
    },
    "id": 7,
    "number": 8,
  },
  {
    "cusp": {
      "zodiac": {
        "id": 4,
        "lord": {
          "id": 0,
          "name": "Sun",
        },
        "name": "Leo",
      },
    },
    "id": 8,
    "number": 9,
  },
  {
    "cusp": {
      "zodiac": {
        "id": 6,
        "lord": {
          "id": 2,
          "name": "Venus",
        },
        "name": "Libra",
      },
    },
    "id": 9,
    "number": 10,
  },
  {
    "cusp": {
      "zodiac": {
        "id": 7,
        "lord": {
          "id": 4,
          "name": "Mars",
        },
        "name": "Scorpio",
      },
    },
    "id": 10,
    "number": 11,
  },
  {
    "cusp": {
      "zodiac": {
        "id": 7,
        "lord": {
          "id": 4,
          "name": "Mars",
        },
        "name": "Scorpio",
      },
    },
    "id": 11,
    "number": 12,
  },
]
`;

exports[`Valid Call calculateBirthChart with valid parameters 3`] = `
[
  {
    "houseNumber": 1,
    "id": 0,
    "isRetrograde": false,
    "name": "Sun",
    "zodiac": {
      "id": 9,
      "lord": {
        "id": 6,
        "name": "Saturn",
      },
      "name": "Capricorn",
    },
  },
  {
    "houseNumber": 11,
    "id": 1,
    "isRetrograde": false,
    "name": "Moon",
    "zodiac": {
      "id": 7,
      "lord": {
        "id": 4,
        "name": "Mars",
      },
      "name": "Scorpio",
    },
  },
  {
    "houseNumber": 1,
    "id": 2,
    "isRetrograde": false,
    "name": "Mercury",
    "zodiac": {
      "id": 9,
      "lord": {
        "id": 6,
        "name": "Saturn",
      },
      "name": "Capricorn",
    },
  },
  {
    "houseNumber": 12,
    "id": 3,
    "isRetrograde": false,
    "name": "Venus",
    "zodiac": {
      "id": 8,
      "lord": {
        "id": 5,
        "name": "Jupiter",
      },
      "name": "Sagittarius",
    },
  },
  {
    "houseNumber": 2,
    "id": 4,
    "isRetrograde": false,
    "name": "Mars",
    "zodiac": {
      "id": 10,
      "lord": {
        "id": 6,
        "name": "Saturn",
      },
      "name": "Aquarius",
    },
  },
  {
    "houseNumber": 4,
    "id": 5,
    "isRetrograde": false,
    "name": "Jupiter",
    "zodiac": {
      "id": 0,
      "lord": {
        "id": 4,
        "name": "Mars",
      },
      "name": "Aries",
    },
  },
  {
    "houseNumber": 5,
    "id": 6,
    "isRetrograde": true,
    "name": "Saturn",
    "zodiac": {
      "id": 1,
      "lord": {
        "id": 2,
        "name": "Venus",
      },
      "name": "Taurus",
    },
  },
  {
    "houseNumber": 2,
    "id": 7,
    "isRetrograde": false,
    "name": "Uranus",
    "zodiac": {
      "id": 10,
      "lord": {
        "id": 6,
        "name": "Saturn",
      },
      "name": "Aquarius",
    },
  },
  {
    "houseNumber": 2,
    "id": 8,
    "isRetrograde": false,
    "name": "Neptune",
    "zodiac": {
      "id": 10,
      "lord": {
        "id": 6,
        "name": "Saturn",
      },
      "name": "Aquarius",
    },
  },
  {
    "houseNumber": 1,
    "id": 9,
    "isRetrograde": false,
    "name": "Pluto",
    "zodiac": {
      "id": 8,
      "lord": {
        "id": 5,
        "name": "Jupiter",
      },
      "name": "Sagittarius",
    },
  },
  {
    "houseNumber": 1,
    "id": 15,
    "isRetrograde": false,
    "name": "Chiron",
    "zodiac": {
      "id": 8,
      "lord": {
        "id": 5,
        "name": "Jupiter",
      },
      "name": "Sagittarius",
    },
  },
  {
    "houseNumber": 1,
    "id": 105,
    "isRetrograde": false,
    "name": "Lilith",
    "zodiac": {
      "id": 8,
      "lord": {
        "id": 5,
        "name": "Jupiter",
      },
      "name": "Sagittarius",
    },
  },
  {
    "houseNumber": 8,
    "id": 103,
    "isRetrograde": true,
    "name": "True North Node",
    "zodiac": {
      "id": 4,
      "lord": {
        "id": 0,
        "name": "Sun",
      },
      "name": "Leo",
    },
  },
  {
    "houseNumber": 8,
    "id": 104,
    "isRetrograde": true,
    "name": "True South Node",
    "zodiac": {
      "id": 4,
      "lord": {
        "id": 0,
        "name": "Sun",
      },
      "name": "Leo",
    },
  },
]
`;

exports[`Invalid Call calculateBirthChart with missing parameters 1`] = `
{
  "error": {
    "issues": [
      {
        "code": "invalid_type",
        "expected": "number",
        "message": "Expected number, received nan",
        "path": [
          "latitude",
        ],
        "received": "nan",
      },
      {
        "code": "invalid_type",
        "expected": "number",
        "message": "Expected number, received nan",
        "path": [
          "longitude",
        ],
        "received": "nan",
      },
    ],
    "name": "ZodError",
  },
  "success": false,
}
`;

exports[`Invalid Call calculateBirthChart with invalid parameters 1`] = `
{
  "error": {
    "issues": [
      {
        "code": "too_big",
        "exact": false,
        "inclusive": true,
        "maximum": 2100,
        "message": "Number must be less than or equal to 2100",
        "path": [
          "year",
        ],
        "type": "number",
      },
      {
        "code": "too_big",
        "exact": false,
        "inclusive": true,
        "maximum": 12,
        "message": "Number must be less than or equal to 12",
        "path": [
          "month",
        ],
        "type": "number",
      },
      {
        "code": "too_small",
        "exact": false,
        "inclusive": true,
        "message": "Number must be greater than or equal to 1",
        "minimum": 1,
        "path": [
          "day",
        ],
        "type": "number",
      },
      {
        "code": "too_big",
        "exact": false,
        "inclusive": true,
        "maximum": 23,
        "message": "Number must be less than or equal to 23",
        "path": [
          "hour",
        ],
        "type": "number",
      },
    ],
    "name": "ZodError",
  },
  "success": false,
}
`;
